Add Zarr benchmarks

oruebel opened the pull request:
Fix #49

This PR makes the following main changes:

- [X] Update `get_object_by_name` to support look-up of an object by name even if a file contains duplicate names (i.e., enforce only that the name of the object we are actually looking for is unique)
- [X] Update `core/_streaming.py` to add functions for reading Zarr NWB files
- [X] Add hdmf-zarr to the requirements
- [X] Add network tracking benchmarks for Zarr (`network_tracking_remote_file_reading.py` and  `network_tracking_remote_slicing.py` )
- [X] Add timing benchmarks for Zarr (`time_remote_file_reading.py` and `time_remote_slicing.py`)
- [x] Update to new `BaseBenchmark` class and parametrize scheme


**Note:** Currently we should use `hdmf-zarr` from the dev branch for this PR since the changes to allow us to configure to not use consolidated metadata have been merged but not released yet (or ideally even use `hdfm-zarr @ git+https://github.com/hdmf-dev/hdmf-zarr.git@fix/consolidated_open_bug`). However, we are planning to a release `hdmf-zarr` with the relevant fixes on Monday so I only listed `hdmf-zarr` in the dependencies, rather than setting a particular branch

oruebel commented:
Note, I ran a few individual tests (which all worked), but  I have not been able to rerun all the benchmarks with the additional tests yet.

CodyCBakerPhD commented on src/nwb_benchmarks/core/_streaming.py:
```diff
@@ -179,3 +181,39 @@ def read_hdf5_nwbfile_ros3(s3_url: str, retry: bool = True) -> Tuple[pynwb.NWBFi
     return (nwbfile, io, retries)
+
+
+def read_zarr(s3_url: str, force_no_consolidated_metadata=False) -> zarr.Group:
```
```suggestion
def read_zarr(s3_url: str, force_no_consolidated_metadata: bool =False) -> zarr.Group:
```

CodyCBakerPhD commented on src/nwb_benchmarks/core/_streaming.py:
```diff
@@ -179,3 +181,39 @@ def read_hdf5_nwbfile_ros3(s3_url: str, retry: bool = True) -> Tuple[pynwb.NWBFi
+    if force_no_consolidated_metadata:
+        zarrfile = zarr.open(store=s3_url, mode="r", storage_options=dict(anon=True))
+    else:
+        try:
+            zarrfile = zarr.open_consolidated(store=s3_url, mode="r", storage_options=dict(anon=True))
+        except KeyError:
+            zarrfile = zarr.open(store=s3_url, mode="r", storage_options=dict(anon=True))
```
Is there no logic we can ask of the Zarr API as to whether there is consolidated metadata available? And if so, I'd adjust the argument here to be `use_consolidated_metadata` which fails with an error if there isn't any

Otherwise, the 'safe' way of just attempting to force it if possible could lead to unclear results in the testing suite since we didn't get feedback as to which route through the logic it actually took on a test case

oruebel replied:
Good point. I have not found anything in Zarr that would allow us to check. The only other check would be to open the file with zarr.open and then check if the .zmetadata attribute exists. Either way there is a check on the remote filesystem needed to see if the .zmetadata file exists. The same logic is being used in hdmf-zarr right now as well. 

That being said, for the benchmarks we should know whether a file does or does not have consolidated metadata (in fact, we probably want to only test with Zarr files that do have it). So maybe we can just always explicitly control in the benchmarks settings whether to use consolidated metadata or not. I.e,. change `force_no_consolidated_metadata` to `open_with_consolidated_metadata` and change this function to:

```
if open_with_consolidated_metadata:
     zarrfile = zarr.open_consolidated(store=s3_url, mode="r", storage_options=dict(anon=True))
else"
        zarrfile = zarr.open(store=s3_url, mode="r", storage_options=dict(anon=True))
```

In this case, if a file does not have consolidated metadata, a test case might fail, but that might be better than benchmarking the wrong thing.

oruebel replied:
I've updated the `read_zarr` method accordingly. If we use a Zarr file without consolidated metadata in benchmarks that assume consolidated metadata, then those benchmarks will fail. But this seems preferable to benchmarking the wrong behavior.

CodyCBakerPhD commented on src/nwb_benchmarks/core/_streaming.py:
```diff
@@ -179,3 +181,39 @@ def read_hdf5_nwbfile_ros3(s3_url: str, retry: bool = True) -> Tuple[pynwb.NWBFi
+    return zarrfile
+
+
+def read_zarr_nwbfile(s3_url: str, force_no_consolidated_metadata=False) -> Tuple[pynwb.NWBFile, hdmf_zarr.NWBZarrIO]:
```
Here it might be clearer to just pass the `mode` and add the details of the `r-` meaning into the docstring?

oruebel replied:
If we want to avoid the try/accept logic for checking if consolidated metadata is available, then we should open the file ourselves in `read_zarr` , because otherwise we run into the same issue in `hdmf_zarr` where that logic actually makes sense, because we don't want the read to just fail there. I'll update the PR and then ping you so you can take a look.

CodyCBakerPhD replied:
Sounds good 👍

oruebel commented:
@CodyCBakerPhD  Please take another look. I've 
- updated `read_zarr` to not do the fallback if consolidated metadata is not present
- updated `read_zarr_nwbfile` to expose the `mode` parameter (NOTE: `hdmf-zarr` does the fallback check under the hood)
- updated the benchmark classes to use the new `BaseBenchmark` class and parametrization scheme

CodyCBakerPhD commented:
Looks good to me 👍
